Sources/SourceKitLSP/Swift/CodeActions/PackageManifestEdits.swift

```diff
@@ -0,0 +1,269 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import LanguageServerProtocol
+import PackageModel
+import PackageModelSyntax
+import SwiftRefactor
+import SwiftSyntax
+
+/// Syntactic code action provider to provide refactoring actions that
+/// edit a package manifest.
+struct PackageManifestEdits: SyntaxCodeActionProvider {
+  static func codeActions(in scope: SyntaxCodeActionScope) -> [CodeAction] {
+    guard let token = scope.firstToken,
+      let call = token.findEnclosingCall()
+    else {
+      return []
+    }
+
+    var actions = [CodeAction]()
+
+    // Add test target(s)
+    actions.append(
+      contentsOf: maybeAddTestTargetActions(call: call, in: scope)
+    )
+
+    // Add product(s).
+    actions.append(
+      contentsOf: maybeAddProductActions(call: call, in: scope)
+    )
+
+    return actions
+  }
+
+  /// Produce code actions to add test target(s) if we are currently on
+  /// a target for which we know how to create a test.
+  static func maybeAddTestTargetActions(
+    call: FunctionCallExprSyntax,
+    in scope: SyntaxCodeActionScope
+  ) -> [CodeAction] {
+    guard let calledMember = call.findMemberAccessCallee(),
+      targetsThatAllowTests.contains(calledMember),
+      let targetName = call.findStringArgument(label: "name")
+    else {
+      return []
+    }
+
+    do {
+      var actions: [CodeAction] = []
+
+      let variants: [(AddTarget.TestHarness, String)] = [
+        (.swiftTesting, "Swift Testing"),
+        (.xctest, "XCTest")
+      ]
+      for (testingLibrary, libraryName) in variants {
+        // Describe the target we are going to create.
+        let target = try TargetDescription(
+          name: "\(targetName)Tests",
+          dependencies: [.byName(name: targetName, condition: nil)],
+          type: .test
+        )
+
+        let edits = try AddTarget.addTarget(
+          target,
+          to: scope.file,
+          configuration: .init(testHarness: testingLibrary)
+        )
+
+        actions.append(
+          CodeAction(
+            title: "Add test target (\(libraryName))",
```

Comment:
A single test target can contain code for both XCTest and swift-testing. (See my comments over on https://github.com/apple/swift-package-manager/pull/7481.)

Reply:
As noted over in https://github.com/apple/swift-package-manager/pull/7481, a code action that creates a brand-new test target that uses two different test libraries is not useful.